File: 5 - Form Template Method/Python/after/loan.py
from after.capital_strategy_advised_line import CapitalStrategyAdvisedLine
from after.capital_strategy_revolver import CapitalStrategyRevolver
from after.capital_strategy_term_loan import CapitalStrategyTermLoan
from after.payment import Payment


class Loan:
    def __init__(self, commitment, outstanding, risk_rating, maturity, expiry, start, today, capital_strategy):
        self.commitment = commitment
        self.outstanding = outstanding
        self.risk_rating = risk_rating
        self.maturity = maturity
        self.expiry = expiry
        self.start = start
        self.today = today
        self.unused_percentage = 1.0
        self.payments = []
        self.capital_strategy = capital_strategy

    def duration(self):
        return self.capital_strategy.duration(self)

    def capital(self):
        return self.capital_strategy.capital(self)

    def payment(self, amount, date):
        self.payments.append(Payment(amount, date))

    def outstanding_risk_amount(self):
        return self.outstanding

    def set_unused_percentage(self, unused_percentage):
        self.unused_percentage = unused_percentage

    def get_unused_percentage(self):
        return self.unused_percentage

    def get_commitment(self):
        return self.commitment

    def get_today(self):
        return self.today

    def get_start(self):
        return self.start

    def get_expiry(self):
        return self.expiry

    def get_payments(self):
        return self.payments

    def get_risk_rating(self):
        return self.risk_rating

    def unused_risk_amount(self):
        return self.commitment - self.outstanding

    @staticmethod
    def new_term_loan(commitment, start, maturity, risk_rating):
        return Loan(commitment, commitment, risk_rating, maturity, None, start, None, CapitalStrategyTermLoan())

    @staticmethod
    def new_revolver(commitment, start, expiry, risk_rating):
        return Loan(commitment, 0.0, risk_rating, None, expiry, start, None, CapitalStrategyRevolver())

    @staticmethod
    def new_advised_line(commitment, start, expiry, risk_rating):
        if risk_rating > 3:
            return None
        advised_line = Loan(commitment, 0.0, risk_rating, None, expiry, start, None, CapitalStrategyAdvisedLine())
        advised_line.set_unused_percentage(0.1)
        return advised_line

File: 4 - Replace Conditional Logic with Strategy/C#/MPG.ReplaceConditionalLogicWithStrategy/MPG.ReplaceConditionalLogicWithStrategy.After/Loan.cs
using System;
using System.Collections.Generic;

// Steps:
// 1. Create class CapitalStrategy
// 2. Declare Capital in CapitalStrategy
// 3. Copy Capital and anything easy from Loan to CapitalStrategy
// 4. Figure out what is needed from a Loan instance, decide next move use context object or data parameters?
// 5. Add getter methods for data that is needed from the context
// 6. Make Loan delegate to CapitalStrategy.Capital
// 7. Move YearsTo, WeightedAverageDuration and Duration to CapitalStrategy
// 8. Make Loan delegate to CapitalStrategy.Duration
// 9. Remove unused code from Loan 
// 10. Extract CapitalStrategy to field
// 11. Extract parameter on field initialization so that CapitalStrategy is passed in via constructor
// 12. Create CapitalStrategyTermLoan, move WeightedAverageDuration to it
// 13. Swap CapitalStrategy for CapitalStrategyTerm loan in factory method
// 14. Delete WeightedAverageDuration from CapitalStrategy as it's only needed for term loans
// 15. Create CapitalStrategyAdvisedLine, move Capital logic, replace in factory method, remove unused code
// 16. Create CapitalStrategyRevolver, move Capital logic, replace in factory method, remove unused code
// 17. Make CapitalStrategy an abstract class, make Capital abstract

namespace MPG.ReplaceConditionalLogicWithStrategy.After
{
    public class Loan
    {
        private readonly double _commitment;
        private readonly int _riskRating;
        private readonly double _outstanding;
        private readonly DateTime? _maturity;
        private readonly DateTime? _expiry;
        private readonly List<Payment> _payments = new List<Payment>();
        private readonly DateTime _start;
        private readonly DateTime? _today;
        private double _unusedPercentage;
        private readonly CapitalStrategy _capitalStrategy;

        public Loan(
            double commitment,
            double outstanding,
            int riskRating,
            DateTime? maturity,
            DateTime? expiry,
            DateTime start,
            DateTime? today,
            CapitalStrategy capitalStrategy)
        {
            _commitment = commitment;
            _riskRating = riskRating;
            _outstanding = outstanding;
            _unusedPercentage = 1.0;
            _maturity = maturity;
            _expiry = expiry;
            _start = start;
            _today = today;
            _capitalStrategy = capitalStrategy;
        }

        public double GetCommitment()
        {
            return _commitment;
        }

        public DateTime? GetExpiry()
        {
            return _expiry;
        }

        public int GetRiskRating()
        {
            return _riskRating;
        }

        public List<Payment> GetPayments()
        {
            return _payments;
        }

        public DateTime? GetToday()
        {
            return _today;
        }

        public DateTime? GetStart()
        {
            return _start;
        }

        internal void SetUnusedPercentage(double unusedPercentage)
        {
            _unusedPercentage = unusedPercentage;
        }

        public double Duration()
        {
            return _capitalStrategy.Duration(this);
        }

        public double Capital()
        {
            return _capitalStrategy.Capital(this);
        }

        public void Payment(double amount, DateTime date)
        {
            _payments.Add(new Payment(amount, date));
        }

        public double OutstandingRiskAmount()
        {
            return _outstanding;
        }

        public double GetUnusedPercentage()
        {
            return _unusedPercentage;
        }

        public double UnusedRiskAmount()
        {
            return _commitment - _outstanding;
        }

        public static Loan NewTermLoan(int commitment, DateTime start, DateTime maturity, int riskRating)
        {
            return new Loan(commitment, commitment, riskRating, maturity, null, start, null, new CapitalStrategyTermLoan());
        }

        public static Loan NewRevolver(double commitment, DateTime start, DateTime expiry, int riskRating)
        {
            return new Loan(commitment, 0, riskRating, null, expiry, start, null, new CapitalStrategyRevolver());
        }

        public static Loan NewAdvisedLine(double commitment, DateTime start, DateTime expiry, int riskRating)
        {
            if (riskRating > 3) return null;
            var advisedLine = new Loan(commitment, 0, riskRating, null, expiry, start, null, new CapitalStrategyAdvisedLine());
            advisedLine.SetUnusedPercentage(0.1);
            return advisedLine;
        }
    }
}
